File: cpp/tests/MergeKSortedListsTest.cpp
#include "MergeKSortedLists.h"
#include "gtest/gtest.h"

using namespace lcpp;

TEST(MergeKSortedLists, Solution23_1) {
  auto S23_1 = Solution23_1();

  std::vector<ListNode *> Lists1{newLinkedList({1, 4, 5}),
                                 newLinkedList({1, 3, 4}),
                                 newLinkedList({2, 6})};
  auto Expected1 = newLinkedList({1, 1, 2, 3, 4, 4, 5, 6});
  auto Actual1 = S23_1.mergeKLists(Lists1);
  EXPECT_EQ(*Expected1, *Actual1);
  deleteLinkedList(Expected1);
  deleteLinkedList(Actual1);

  std::vector<ListNode *> Lists2;
  auto Expected2 = newLinkedList({});
  auto Actual2 = S23_1.mergeKLists(Lists2);
  EXPECT_EQ(*Expected2, *Actual2);
  deleteLinkedList(Expected2);
  deleteLinkedList(Actual2);

  std::vector<ListNode *> Lists3{newLinkedList({})};
  auto Expected3 = newLinkedList({});
  auto Actual3 = S23_1.mergeKLists(Lists3);
  EXPECT_EQ(*Expected3, *Actual3);
  deleteLinkedList(Expected3);
  deleteLinkedList(Actual3);
}

TEST(MergeKSortedLists, Solution23_2) {
  auto S23_2 = Solution23_2();

  std::vector<ListNode *> Lists1{newLinkedList({1, 4, 5}),
                                 newLinkedList({1, 3, 4}),
                                 newLinkedList({2, 6})};
  auto Expected1 = newLinkedList({1, 1, 2, 3, 4, 4, 5, 6});
  auto Actual1 = S23_2.mergeKLists(Lists1);
  EXPECT_EQ(*Expected1, *Actual1);
  deleteLinkedList(Expected1);
  deleteLinkedList(Actual1);

  std::vector<ListNode *> Lists2;
  auto Actual2 = S23_2.mergeKLists(Lists2);
  EXPECT_EQ(nullptr, Actual2);

  std::vector<ListNode *> Lists3{nullptr};
  auto Actual3 = S23_2.mergeKLists(Lists3);
  EXPECT_EQ(nullptr, Actual3);
}

File: java/src/test/java/io/binac/leetcode/MergeKSortedListsTest.java
package io.binac.leetcode;

import io.binac.leetcode.util.LinkedLists;
import io.binac.leetcode.util.ListNode;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class MergeKSortedListsTest {
    private static final MergeKSortedLists.Solution1 SOLUTION1 = new MergeKSortedLists.Solution1();
    private static final MergeKSortedLists.Solution2 SOLUTION2 = new MergeKSortedLists.Solution2();

    private final ListNode[] LISTS1 = {
            LinkedLists.asLinkedList(1, 4, 5),
            LinkedLists.asLinkedList(1, 3, 4),
            LinkedLists.asLinkedList(2, 6)
    };
    private final ListNode EXPECTED1 = LinkedLists.asLinkedList(1, 1, 2, 3, 4, 4, 5, 6);

    private final ListNode[] LISTS2 = {};
    private final ListNode EXPECTED2 = null;

    private final ListNode[] LISTS3 = {LinkedLists.asLinkedList()};
    private final ListNode EXPECTED3 = null;


    @Test
    void test1() {
        assertTrue(LinkedLists.equals(EXPECTED1, SOLUTION1.mergeKLists(LISTS1)));
        assertTrue(LinkedLists.equals(EXPECTED2, SOLUTION1.mergeKLists(LISTS2)));
        assertTrue(LinkedLists.equals(EXPECTED3, SOLUTION1.mergeKLists(LISTS3)));
    }

    @Test
    void test2() {
        assertTrue(LinkedLists.equals(EXPECTED1, SOLUTION2.mergeKLists(LISTS1)));
        assertTrue(LinkedLists.equals(EXPECTED2, SOLUTION2.mergeKLists(LISTS2)));
        assertTrue(LinkedLists.equals(EXPECTED3, SOLUTION2.mergeKLists(LISTS3)));
    }
}